### ISSUE TITLE
How to get any kind of notification once the user clicks on "Don't Allow" to camera persimmon alert view.

### ISSUE THREAD

**Issue by jshivam** [twilio/twilio-video-ios#165]
### Description

I can't find any api exposed by Twillio where i can get any callback once the user clicks on "Don't Allow" to camera persimmon alert view.

How to get any kind of notification once the user doesn't allow the camera permission ?

**Comment by paynerc** [comment]
@jshivam,

The Camera and Microphone permissions are handled not by Twilio but by Apple via `AVFoundation`. Take a look at [`AVCaptureDevice authorizationStatusForMediaType:`](https://developer.apple.com/documentation/avfoundation/avcapturedevice/1624613-authorizationstatusformediatype?language=objc).

In your app you can do something similar:

```
switch ([AVCaptureDevice authorizationStatusForMediaType:AVMediaTypeVideo]) {
        case AVAuthorizationStatusNotDetermined: {
            // If the status is not determined, it means the user has never been prompted in your app for camera permissions, so you can trigger the camera prompt at a time more convenient for your app flow:
            [AVCaptureDevice requestAccessForMediaType:AVMediaTypeVideo completionHandler:^(BOOL granted) {
                NSLog(@"Camera permissions granted? %@", granted ? @"YES" : @"NO");
            }];
            break;
        }
        case AVAuthorizationStatusAuthorized: {
            // This means the user was prompted previously and they granted permissions, normal work flow can happen here
            NSLog(@"Camera permissions were previously granted.");
            break;
        }
        default: {
            // This means that the user previously declined camera permissions. Here you could provide an alert stating why you need camera permissions and direct them to your app's settings page in the `Settings.app` and suggest that they enable camera permissions if they wish to access this functionality.
            NSLog(@"Camera permissions were previously declined.");
        }
    }
```

Handling microphone permissions is similar, just change the `mediaType` parameter from `AVMediaTypeVideo` to `AVMediaTypeAudio`.

Let me know if this helps.

Ryan

**Comment by jshivam** [comment]
```swift
        camera.startCapture(device: frontCamera ?? backCamera) { [weak self] (captureDevice, videoFormat, error) in
            guard let `self` = self else { return }
            if error.isNotNil {
                self.enableCameraMode(false)
            } else {
                self.videoView.shouldMirror = (captureDevice.position == .front)
                self.enableCameraMode(true)
            }
        }
```
@paynerc paynerc
When I call `camera.startCapture` for the first time, it ask for camera persimmon, but when I deny it, it doesn't give me any error in the closure.  I believe it should give me an error in the closure. But it does't. 

Note: it gives me the error in the closure when I run the code for the second time. i.e. when it's known that `AVAuthorizationStatus` is either `AVAuthorizationStatusDenied` or `AVAuthorizationStatusRestricted`

**Comment by paynerc** [comment]
@jshivam,

Ah, I see. You are calling `startCapture` and it is returning the closure as it begins to initialize the capturer. However, if you don't allow the camera at that point, you aren't getting any notification that the camera permission was declined. I will take a look to see why a call back isn't raised to let you know the camera capture should be stopped due to lack of permissions.

Ryan

**Comment by paynerc** [comment]
@jshivam,

From what I am seeing, the first call to `AVCaptureDeviceInput initWithDevice:error:` is what will trigger the Camera permissions dialog to be raised if the permissions have not been previously granted. Doing some further investigation, it turns out that the `AVCaptureSession` is a bit weird on the first run. If you decline camera access, it will then return you black frames during that initial session. We are not getting any sort of notification from the session to alert us that the camera permissions were denied so it looks to us that everything is fine. And as you mentioned, on a subsequent `startCapture` request, it will indeed fail with the error `Cannot use Front Camera`. So there is definitely a corner case on the first usage of the camera in an application whereby the user declines the camera permission.

One thing that can be done in your application is to take the approach I mentioned earlier. At some point in your application prior to attempting to use the camera, you check the camera permission, trigger the camera prompt if it is undetermined, and if it was declined previously, display a message to the user to visit the Settings.app to enable the camera so they can capture video.

I am investigating what we could do internally to check for the camera permissions earlier in the `startCapture` flow and see if we could trigger the prompt and if permission was then denied, be able to return an error on `startCapture` instead of having it go forward with the black frames.

Ryan

**Comment by paynerc** [comment]
@jshivam,

We released Twilio Video iOS `4.3.0` which now properly returns an error in the `CameraSource.startCapture()` completion if the user has declined the camera permissions. Please test out your scenario where the user declines camera permissions on the first time permissions are requested when calling `CameraSource.startCapture()`.

Ryan

**Comment by paynerc** [comment]
@jshivam,

Just checking in to see if this has solved your issue. I will close this issue, but feel free to reopen it if you are still experiencing issues.

Ryan
